File: 15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#
from typing import List
# @lc code=start
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        answer = []
        nums.sort()
        
        # X + Y = -Z
        for i in range(len(nums)):
            if i > 0 and nums[i-1] == nums[i]:
                # 이전에 숫자와 같은 경우 제외 (중복 답)
                continue
            Z = -nums[i] # target Value
            
            a = i + 1 # 왼쪽부터 차례대로 다 하니까, i값 기준으로 오른쪽만 하면 됨
            b = len(nums) - 1

            while a < b:
                X = nums[a]
                Y = nums[b]
            
                if X + Y == Z:
                    answer.append([X, Y, -Z])
                    while a < b and nums[a] == nums[a+1]:
                        a += 1
                    while a < b and nums[b] == nums[b-1]:
                        b -= 1
                    a += 1
                    b -= 1
                    # a, b 각 숫자를 전과 동일하지 않게 함
                elif X + Y < Z:
                    a += 1
                elif X + Y > Z:
                    b -= 1
        
        return answer

# @lc code=end

print(Solution.threeSum(None, [-1,0,1,2,-1,-4]))

File: 15.3-sum.js
/*
 * @lc app=leetcode id=15 lang=javascript
 *
 * [15] 3Sum
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
  const answer = [];
  nums.sort((a, b) => a - b);

  // Z = X + Y
  for (var i = 0; i < nums.length; i++) {
    if (i > 0 && nums[i] === nums[i - 1]) {
      // 중복 경우 제외를 위해 같은 숫자는 패스
      continue;
    }
    
    var Z = -nums[i];

    var left = i + 1;
    var right = nums.length - 1;

    while (left < right) {
      var X = nums[left];
      var Y = nums[right];

      if (X + Y === Z) {
        answer.push([-Z, X, Y]);

        // 중복 숫자들 모두 넘어감
        while (left < right && nums[left] === nums[left + 1]) {
          left++;
        }

        while (left < right && nums[right] === nums[right - 1]) {
          right--;
        }

        left++;
        right--;
      } else if (X + Y < Z) {
        // 더 큰 숫자를 더해야 Z가 됨 -> 작은 숫자 쪽에 있는 left를 더한다
        left++;
      } else if (X + Y > Z) {
        // 더 작은 숫자를 더해야 함 -> 큰 숫자 쪽의 right를 줄인다
        right--;
      }
    }
  }
  return answer;  
};
// @lc code=end
